syntax = "proto3";

package sum;

service SumService {
  // vectors CRUD
  rpc CreateRecord(Record) returns (RecordResponse) {}
  rpc UpdateRecord(Record) returns (RecordResponse) {}
  rpc ReadRecord(ById) returns (RecordResponse) {}
  rpc DeleteRecord(ById) returns (RecordResponse) {}
  
  // oracles CRUD
  rpc CreateOracle(Oracle) returns (OracleResponse) {}
  rpc UpdateOracle(Oracle) returns (OracleResponse) {}
  rpc ReadOracle(ById) returns (OracleResponse) {}
  rpc FindOracle(ByName) returns (OracleResponse) {}
  rpc DeleteOracle(ById) returns (OracleResponse) {}

  // core methods
  // TODO: rpc Eval(EvalRequest) (???) {}
  rpc Info(Empty) returns (ServerInfo) {}
}

message NamedValue {
    string name = 1;
    string value =  2;
}

message Record {
    string id = 1;
    repeated float data = 2;
    repeated NamedValue meta = 3;
}

message RecordResponse {
    bool success = 1;
    string msg = 2;
    Record record = 3;
}

message Oracle {
    string id = 1;
    string name = 2;
    string code = 3;
}

message OracleResponse {
    bool success = 1;
    string msg = 2;
    repeated Oracle oracles = 3;
}

message EvalRequest {
    string oracle_id = 1;
    repeated NamedValue args = 2;
}

message ById {
    string id = 1;
}

message ByName {
    string name = 1;
}

message ServerInfo {
    string version = 1;
    uint64 uptime  = 2;
    uint64 pid = 3;
    uint64 uid = 4;
    repeated string argv = 5;
    uint64 records = 6;
    uint64 oracles = 7;
}

message Empty {}
